lib/types.js

```diff
@@ -1246,38 +1301,33 @@ UnionType.prototype._branchConstructor = function () {
  * + `map`, `record`
  */
 class UnwrappedUnionType extends UnionType {
-  constructor (schema, opts) {
+  /**
+   *
+   * @param {*} schema
+   * @param {*} opts
+   * @param {Function|undefined} projectionFn The projection function used
+   *                                          to determine the bucket for the
+   *                                          Union. Falls back to generate
+   *                                          from `wrapUnions` parameter
+   *                                          if given.
+   */
+  constructor (schema, opts, projectionFn) {
     super(schema, opts);
 
-    this._dynamicBranches = null;
-    this._bucketIndices = {};
-    this.types.forEach(function (type, index) {
-      if (Type.isType(type, 'abstract', 'logical')) {
-        if (!this._dynamicBranches) {
-          this._dynamicBranches = [];
-        }
-        this._dynamicBranches.push({index, type});
-      } else {
-        let bucket = getTypeBucket(type);
-        if (this._bucketIndices[bucket] !== undefined) {
-          throw new Error(`ambiguous unwrapped union: ${j(this)}`);
-        }
-        this._bucketIndices[bucket] = index;
+    if (!projectionFn && opts && typeof opts.wrapUnions === 'function') {
+      try {
+        projectionFn = opts.wrapUnions(this.types);
+      } catch(e) {
+        throw new Error(`Error generating projection function: ${e}`);
       }
-    }, this);
+    }
+    this._getIndex = projectionFn
+      ? generateProjectionIndexer(projectionFn)
+      : generateDefaultIndexer.bind(this)(this.types);
 
     Object.freeze(this);
   }
 
-  _getIndex (val) {
-    let index = this._bucketIndices[getValueBucket(val)];
-    if (this._dynamicBranches) {
-      // Slower path, we must run the value through all branches.
-      index = this._getBranchIndex(val, index);
-    }
-    return index;
-  }
-
   _getBranchIndex (any, index) {
```

Comment:
we could move this out as well or alternatively as described above, move the `defaultIndexer` into the class.

Reply:
Yep, let's move it out. Carving unnecessary logic out makes `UnwrappedUnion` simpler and bucket inference more modular.

Reply:
bd2c6dd